x-pack/plugins/fleet/server/tasks/unenroll_inactive_agents_task.ts

```diff
@@ -0,0 +1,191 @@
+export const VERSION = '1.0.0';
+const TITLE = 'Fleet Deleted Files Periodic Tasks';
+const SCOPE = ['fleet'];
+const INTERVAL = '10m';
+const TIMEOUT = '1m';
+const POLICIES_BATCHSIZE = 100;
+const UNENROLLMENT_BATCHSIZE = 1000;
+
+interface UnenrollInactiveAgentsTaskSetupContract {
+  core: CoreSetup;
+  taskManager: TaskManagerSetupContract;
+  logFactory: LoggerFactory;
+}
+
+interface UnenrollInactiveAgentsTaskStartContract {
+  taskManager: TaskManagerStartContract;
+}
+
+export class UnenrollInactiveAgentsTask {
+  private logger: Logger;
+  private wasStarted: boolean = false;
+  private abortController = new AbortController();
+
+  constructor(setupContract: UnenrollInactiveAgentsTaskSetupContract) {
+    const { core, taskManager, logFactory } = setupContract;
+    this.logger = logFactory.get(this.taskId);
+
+    taskManager.registerTaskDefinitions({
+      [TYPE]: {
+        title: TITLE,
+        timeout: TIMEOUT,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              return this.runTask(taskInstance, core);
+            },
+            cancel: async () => {
+              this.abortController.abort('Task timed out');
+            },
+          };
+        },
+      },
+    });
+  }
+
+  public start = async ({ taskManager }: UnenrollInactiveAgentsTaskStartContract) => {
+    if (!taskManager) {
+      this.logger.error('[UnenrollInactiveAgentsTask] Missing required service during start');
+      return;
+    }
+
+    this.wasStarted = true;
+    this.logger.info(`[UnenrollInactiveAgentsTask] Started with interval of [${INTERVAL}]`);
+
+    try {
+      await taskManager.ensureScheduled({
+        id: this.taskId,
+        taskType: TYPE,
+        scope: SCOPE,
+        schedule: {
+          interval: INTERVAL,
+        },
+        state: {},
+        params: { version: VERSION },
+      });
+    } catch (e) {
+      this.logger.error(`Error scheduling task UnenrollInactiveAgentsTask, error: ${e.message}`, e);
+    }
+  };
+
+  private get taskId(): string {
+    return `${TYPE}:${VERSION}`;
+  }
+
+  private endRun(msg: string = '') {
+    this.logger.info(`[UnenrollInactiveAgentsTask] runTask ended${msg ? ': ' + msg : ''}`);
+  }
+
+  public async unenrollInactiveAgents(
+    esClient: ElasticsearchClient,
+    soClient: SavedObjectsClientContract
+  ) {
+    this.logger.debug(
+      `[UnenrollInactiveAgentsTask] Fetching agent policies with unenroll_timeout > 0`
+    );
+    // find all agent policies that are not managed and having unenroll_timeout > 0
+    // limit batch size to 100 to avoid scale issues
+    const policiesKuery = `${AGENT_POLICY_SAVED_OBJECT_TYPE}.is_managed: false AND ${AGENT_POLICY_SAVED_OBJECT_TYPE}.unenroll_timeout > 0`;
+    const agentPolicies = await agentPolicyService.list(soClient, {
+      kuery: policiesKuery,
+      perPage: POLICIES_BATCHSIZE,
```

Comment:
The issue with this approach is if there are > 100 policies with unenroll timeout, the task will never unenroll agents enrolled to the policies after the first 100. Perhaps we should add a for loop to iterate all policies, and break out if 1k agents were unenrolled in this run of the task.

Reply:
yes I think you're right. I could increase the POLICIES_BATCHSIZE to a bigger value (500 for instance) but we'd still need to loop over the remaining policies.

Reply:
In https://github.com/elastic/kibana/pull/189861/commits/dc397abe44f6659243f2cc83bd2073a012333622 I switched `POLICIES_BATCHSIZE` to `SO_SEARCH_LIMIT`, so we retrieve all the policies. Then the agents query is limited to 1000, so if the following are not unenrolled on the first run will be done in the next batch after 10 minutes.

Reply:
I think `SO_SEARCH_LIMIT` might not work to create an agent query on `policy_id` with `or` condition below, and it has a risk of loading 10k policies in the memory at once. Using a for loop would probably be safer.

Reply:
@juliaElastic I pushed another commit, now the agent policy query is inside a loop and each batch is again of 500. It goes on until there are no more agent policies, it's very similar to what we already do [here](https://github.com/elastic/kibana/blob/5652080dbb665c32128397e7e9196f4536ccbe2c/x-pack/plugins/fleet/server/services/agents/update.ts#L15)